app/controllers/doorkeeper/tokens_controller.rb

```diff
@@ -88,7 +88,20 @@ def strategy
     end
 
     def authorize_response
-      @authorize_response ||= strategy.authorize
+      @authorize_response ||= begin
+        before_successful_authorization
+        auth = strategy.authorize
+        after_successful_authorization if auth.is_a?(Doorkeeper::OAuth::TokenResponse)
```

Comment:
Can you please point me why we call this hook only for `TokenResponse` ? Not `CodeResponse`, for example

**UPD**: ahh, seems like because of your question about pre authorization, right?

Reply:
Yes. That (pre authorization) and my meagre understanding of how it's supposed to work. 😄 

Initially I had this as:

```Ruby
after_successful_authorization unless auth.is_a?(Doorkeeper::OAuth::ErrorResponse)
```

☝️ would that work?

PS. I just noticed that `Doorkeeper::OAuth::TokenResponse` in not inheriting from `BaseResponse` like `Doorkeeper::OAuth::CodeResponse`. 😱

Reply:
Actually I don't remember. Need some time to check it :)

Reply:
I changed it to what I proposed above which makes more sense anyway, since there might be a `Doorkeeper::OAuth::CodeResponse` as well (for `grant_type=authorization_code`).

I'd ❤️ some feedback for the `pre_auth` part! 🙏